metricflow/cli/dbt_connectors/adapter_backed_client.py

```diff
@@ -0,0 +1,308 @@
+    ) -> None:
+        """Dry run statement; checks that the 'stmt' is queryable. Returns None on success.
+
+        Raises an exception if the 'stmt' isn't queryable.
+
+        Args:
+            stmt: The SQL query statement to dry run.
+            sql_bind_parameters: The parameter replacement mapping for filling in
+                concrete values for SQL query parameters.
+        """
+        start = time.time()
+        logger.info(
+            f"Running dry_run of:"
+            f"\n\n{indent_log_line(stmt)}\n"
+            + (f"\nwith parameters: {dict(sql_bind_parameters.param_dict)}" if sql_bind_parameters.param_dict else "")
+        )
+        # TODO - rely on self._adapter.dry_run() when it is available so this will work for BigQuery.
+        self.execute(f"EXPLAIN {stmt}")
+        stop = time.time()
+        logger.info(f"Finished running the dry_run in {stop - start:.2f}s")
+        return
+
+    def create_table_from_dataframe(
+        self,
+        sql_table: SqlTable,
+        df: pd.DataFrame,
+        chunk_size: Optional[int] = None,
+    ) -> None:
+        """Create a table in the data warehouse containing the contents of the dataframe.
+
+        Only used in tutorials and tests.
+
+        Args:
+            sql_table: The SqlTable object representing the table location to use
+            df: The Pandas DataFrame object containing the column schema and data to load
+            chunk_size: The number of rows to insert per transaction
+        """
+        logger.info(f"Creating table '{sql_table.sql}' from a DataFrame with {df.shape[0]} row(s)")
+        start_time = time.time()
+        with self._adapter.connection_named("MetricFlow_create_from_dataframe"):
+            # Create table
+            # update dtypes to convert None to NA in boolean columns.
+            # This mirrors the SQLAlchemy schema detection logic in pandas.io.sql
+            df = df.convert_dtypes()
+            columns = df.columns
+            columns_to_insert = []
+            for i in range(len(df.columns)):
+                # Format as "column_name column_type"
+                columns_to_insert.append(
+                    f"{columns[i]} {self._get_type_from_pandas_dtype(str(df[columns[i]].dtype).lower())}"
+                )
+            self._adapter.execute(
+                f"CREATE TABLE IF NOT EXISTS {sql_table.sql} ({', '.join(columns_to_insert)})",
+                auto_begin=True,
+                fetch=False,
+            )
+            self._adapter.commit_if_has_connection()
+
+            # Insert rows
+            values = []
+            for row in df.itertuples(index=False, name=None):
+                cells = []
+                for cell in row:
+                    if pd.isnull(cell):
+                        # use null keyword instead of isNA/None/etc.
+                        cells.append("null")
+                    elif type(cell) in [str, pd.Timestamp]:
```

Comment:
Generally works, but I vaguely recall running into type errors with one SQL engine where specifying a string when a time is expected threw an error. Cross that bridge if / when we get there.